sherpa/astro/io/__init__.py

```diff
@@ -167,7 +163,7 @@ def read_arrays(*args):
         raise IOErr('noarrays')
 
     dstype = Data1D
-    if _is_subclass(args[-1], BaseData):
+    if sherpa.io._is_subclass(args[-1], BaseData):
```

Comment:
I believe 
```suggestion
    if is_subclass(args[-1], BaseData):
```
would be sufficient. I mean, is it really the task of this function to check that it's not `BaseData` itself? If args[-1] is BaseData and it work, fine, if not, BaseData will throw an exception. But this is fine, too.

Reply:
Going forward to your "trying not to touch code ..." comment, I have not looked at our various 
is an instance of" routines and compared them to standard routines. In this case it's

```python
def _is_subclass(t1, t2):
    return isinstance(t1, type) and issubclass(t1, t2) and (t1 is not t2)
```

so it does something more than `issubclass`. Although **why** it does these - surely the `isinstance` check is essentially a no-op (this comes from the old days, written in Python 2 times, so maybe it had some use earlier). In fact, it looks like the important thing is the `t1 is not t2` check. We don't want people to be sending in `BaseData`, but to be honest if they do so then things aren't going to work out well for them anyway.

@olaurino left `BaseData` in when he re-worked the data hierarchy in #614 - leaving behind the following

```python
class BaseData(metaclass=ABCMeta):
    """
    Base class for all data classes. Left for compatibility with older versions.
    """
    pass
```

So I think here we are okay in changing to `issubclass`. Maybe it's time to remove `BaseData`, but that would be a different PR ;-)

As an aside, `sherpa.ui.utils` has it's own variant

```python
def _is_subclass(t1, t2):
    return inspect.isclass(t1) and issubclass(t1, t2) and (t1 is not t2)
```

so an alternative for this PR would be to remove `sherpa.io._is_subclass` and replace by `sherpa.ui.utils._is_subclass` (since this PR was trying to reduce the number of times we re-write the same code), but maybe we should just use `is_subclass` there too (but again in another PR).

Reply:
Hmm. Actually I think there are a number of CIAO and/or L3 scripts that could rely on `BaseData`, so there would be a cost to changing this. But again, for a different PR.

Reply:
Right, I agree that is does more - but I think that "more" does not have to be done *here* but, I agree, not for this PR.